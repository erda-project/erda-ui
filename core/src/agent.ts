// Copyright (c) 2021 Terminus, Inc.
//
// This program is free software: you can use, redistribute, and/or modify
// it under the terms of the GNU Affero General Public License, version 3
// or later ("AGPL"), as published by the Free Software Foundation.
//
// This program is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program. If not, see <http://www.gnu.org/licenses/>.

import agentUse from 'superagent-use';
import { getCurrentLocale } from './i18n';
import request from 'superagent';

const superagent = agentUse(request);

function getCookies(key: string) {
  const cookies = {};
  window.document.cookie.split(';').forEach((item) => {
    const [k, v] = item.split('=');
    cookies[k.trim()] = v && v.trim();
  });
  return key ? cookies[key] : cookies;
}

/**
 * set accept header
 */
function setHeader(req: request.SuperAgentRequest) {
  req.set('Accept', 'application/vnd.dice+json;version=1.0');
  req.set('Lang', getCurrentLocale().key === 'zh' ? 'zh-CN' : 'en-US');
  if (!['GET', 'HEAD', 'OPTIONS', 'TRACE'].includes(req.method)) {
    const token = getCookies('OPENAPI-CSRF-TOKEN');
    if (token) {
      req.set('OPENAPI-CSRF-TOKEN', token);
    }
  }
  return req;
}

superagent.use(setHeader);

export default superagent;
