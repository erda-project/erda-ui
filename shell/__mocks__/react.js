// Copyright (c) 2021 Terminus, Inc.
//
// This program is free software: you can use, redistribute, and/or modify
// it under the terms of the GNU Affero General Public License, version 3
// or later ("AGPL"), as published by the Free Software Foundation.
//
// This program is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program. If not, see <http://www.gnu.org/licenses/>.

const react = require('react');
// Resolution for requestAnimationFrame not supported in jest error :
// https://github.com/facebook/react/issues/9102#issuecomment-283873039
global.window = global;
window.addEventListener = () => {};
window.requestAnimationFrame = () => {
  // throw new Error('requestAnimationFrame is not supported in Node');
};

module.exports = react;
