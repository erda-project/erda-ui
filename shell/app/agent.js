// Copyright (c) 2021 Terminus, Inc.
//
// This program is free software: you can use, redistribute, and/or modify
// it under the terms of the GNU Affero General Public License, version 3
// or later ("AGPL"), as published by the Free Software Foundation.
//
// This program is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program. If not, see <http://www.gnu.org/licenses/>.

import agent from 'core/agent';
import { getCookies, setApiWithOrg, getOrgFromPath } from 'common/utils';
import { getGlobal } from 'app/global-space';
import { some } from 'lodash';
import errorHandler from './error-handler';


const isExcludeOrgHeaderApi = (url) => {
  const excludeApis = ['/api/files'];
  return some(excludeApis, api => api.startsWith(url));
};

function handleUrl(req) {
  const header = getGlobal('service-provider');
  if (header) {
    req.set('service-provider', header);
  }
  const { url } = req;

  // 处理api/spot接口前缀
  if (url.startsWith('/api/spot/')) {
    req.url = url.replace('/api/spot/', '/api/');
  }

  const curOrg = getOrgFromPath();
  curOrg && req.set('org', curOrg);
  if (!isExcludeOrgHeaderApi(req.url)) {
    req.url = setApiWithOrg(req.url);
  }
}

function handleError(req) {
  req.retry(2, (err, response) => {
    // csrf token过期时进行重试
    if (err && err.status === 403 && (response.text || '').includes('empty csrf token')) {
      // 重试的肯定是需要带token的，不必再判断method
      const token = getCookies('OPENAPI-CSRF-TOKEN');
      if (token) {
        req.set('OPENAPI-CSRF-TOKEN', token);
      }
      return true;
    }
    return false;
  });
  req.on('error', errorHandler);
}

// if paging list is null, transform to array
function handelPagingNull(req) {
  req.on('response', (res) => {
    if (res.body && res.body.data) {
      const { data } = res.body;
      if (
        typeof data === 'object' &&
        data !== null &&
        'list' in data &&
        'total' in data &&
        data.list === null
      ) {
        data.list = [];
      }
    }
  });
}

const reqMap = {};
const reqKeyList = [];
// const fetchingReqMap = {};
/**
 * mark & save request for later abort
 * multi same url requests only preserve one
 */
// function mark(req) {
// const key = null;

// req.on('request', () => {
//   key = reqKeyList.pop();
//   const url = req.url;
//   // if same url request is fetching, abort others just left one
//   if (fetchingReqMap[url] === undefined) {
//     fetchingReqMap[url] = req;
//   } else {
//     // it seems directly call req.abort() not work,so abort previous request
//     fetchingReqMap[url].abort();
//     fetchingReqMap[url] = req;
//   }
//   if (reqKeyList.length) {
//     console.warn('request key list should be empty, but left: ', [...reqKeyList]);
//     reqKeyList.length = 0;
//   }
//   reqMap[key] = req;
// });
// req.on('response', () => {
//   delete reqMap[key];
//   delete fetchingReqMap[req.url];
// });
// return req;
// }

// function markReq(key) {
//   reqKeyList.push(key);
// }

// function abortReq(key) {
//   if (Array.isArray(key)) {
//     key.forEach((k) => {
//       reqMap[k] && reqMap[k].abort();
//     });
//   } else {
//     reqMap[key] && reqMap[key].abort();
//   }
// }

// mf_share导出的模块内部会引用这里的agent，导致use方法被执行两次
agent.use(req => {
  if (!req._marked) {
    req._marked = true;
    handleUrl(req);
    handelPagingNull(req);
    handleError(req);
  }
});
export default agent;
