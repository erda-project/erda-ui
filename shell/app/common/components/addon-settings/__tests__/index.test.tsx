// Copyright (c) 2021 Terminus, Inc.
//
// This program is free software: you can use, redistribute, and/or modify
// it under the terms of the GNU Affero General Public License, version 3
// or later ("AGPL"), as published by the Free Software Foundation.
//
// This program is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program. If not, see <http://www.gnu.org/licenses/>.

import React from 'react';
import AddonSettings from '..';
import routeInfoStore from 'core/stores/route';
import { render } from '@testing-library/react';
import agent from 'agent';

const { PureAddonSettings } = AddonSettings;

const insId = 123;
const routerData = {
  params: {
    insId,
  },
};
const data = {
  name: 'erda',
};

jest.mock('core/stores/route');
jest.mock('agent');

describe('addon-settings', () => {
  const fn = jest.fn();
  beforeAll(() => {
    routeInfoStore.getState = (fn) => {
      return fn(routerData);
    };
    routeInfoStore.useStore = (fn) => {
      return fn(routerData);
    };
    agent.get = jest.fn().mockResolvedValue({
      body: {
        success: true,
        data,
      },
    });
    agent.delete = () => {
      fn.call(null);
      return Promise.resolve({
        body: {
          success: true,
          data,
        },
      });
    };
  });
  afterAll(() => {
    jest.clearAllMocks();
  });
  describe.only('AddonSettings', () => {
    it('should render well', async () => {
      const wrapper = render(<AddonSettings />);
      expect(wrapper.container).toMatchSnapshot();
      // expect(wrapper.find('PureAddonSettings').prop('addonConfig')).toStrictEqual(data);
      // expect(wrapper.find('PureAddonSettings').prop('insId')).toBe(insId);
    });
  });
  describe('PureAddonSettings', () => {
    const config = {
      name: 'erda',
      org: 'erda.cloud',
    };
    it('should render well', async () => {
      const wrapper = mount(<PureAddonSettings insId={insId} />);
      expect(wrapper).toBeEmptyRender();
      wrapper.setProps({
        addonConfig: {
          config,
          canDel: false,
        },
      });
      wrapper.update();
      expect(wrapper.find('.param-k')).toHaveLength(2);
      expect(wrapper.find('.settings-delete')).not.toExist();
      wrapper.setProps({
        addonConfig: {
          config,
          canDel: true,
        },
      });
      wrapper.update();
      expect(wrapper.find('.settings-delete')).toExist();
      await wrapper.find('[deleteItem="service"]').prop('onConfirm')();
      expect(fn).toHaveBeenCalledTimes(1);
    });
  });
});
