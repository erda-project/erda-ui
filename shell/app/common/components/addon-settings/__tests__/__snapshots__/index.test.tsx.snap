// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addon-settings AddonSettings should render well 1`] = `<div />`;
